### PR TITLE
patch: ignore -i in requirements.txt

### LINKED ISSUE #1406: Pipenv lock requirements.txt failes validation with InvalidRequirement
**Describe the bug**
I am using `Pipenv` to manage package environments, and so I want to use it to generate a requirements file for use in the bento package. However, the file generated by `Pipenv` fails validation when calling `.save()` or `.save_to_dir()`.

```
InvalidRequirement: Parse error at "'-i https'": Expected W:(abcd...)
```

To generate a `requirements.txt` file for non-development packages with `pipenv` one can run the command `pipenv lock -r > requirements.txt`.

The top of that file looks like this
```
#
# These requirements were autogenerated by pipenv
# To regenerate from the project's Pipfile, run:
#
#    pipenv lock --requirements
#

-i https://pypi.org/simple
```

It looks like the like beginning `-i` is not expected by the validation mechanism in bentoml, even though it's a valid requirements file for pip.

**To Reproduce**
Using `pipenv` to manage a python environment,  create a `requirements.txt` using the command
```
pipenv lock -r > requirements.txt
```

Then refer to this file in the bento definition like so
```
@bentoml.env(requirements_txt_file="./requirements.txt")
```

When saving the bento, you will see the error mentioned above.

**Expected behavior**
Bentoml accepts this format of requirements.txt without any changes.

**Environment:**
 - OS: MacOS 10.15.7
 - Python 3.8.6
 - pipenv, version 2020.11.15
 - bentoml, version 0.10.1

**Additional context**
A workaround for now is to remove the offending line when creating the `requirements.txt` file with something like this:
```
pipenv lock -r > requirements.txt && sed -i -e "/^-i/d" requirements.txt && rm requirements.txt-e
```

### PR DESCRIPTION
**Pull request by jackyzha0**
## Description
pip supports `requirements.txt` with customized PyPI URL which is prefixed with a `-i` flag. We add this support by changing `requirements_txt_file` option to copy rather than parse.

## Motivation and Context
closes #1406 

## Types of changes
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature and improvements (non-breaking change which adds/improves functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Code Refactoring (internal change which is not user facing)
- [ ] Documentation
- [ ] Test, CI, or build

## Component(s) if applicable
- [x] BentoService (service definition, dependency management, API input/output adapters)
- [ ] Model Artifact (model serialization, multi-framework support)
- [ ] Model Server (mico-batching, dockerisation, logging, OpenAPI, instruments)
- [ ] YataiService gRPC server (model registry, cloud deployment automation)
- [ ] YataiService web server (nodejs HTTP server and web UI)
- [ ] Internal (BentoML's own configuration, logging, utility, exception handling)
- [ ] BentoML CLI

## Checklist:
- [x] My code follows the bentoml code style, both `./dev/format.sh` and
  `./dev/lint.sh` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [x] My change reduces project test coverage and requires unit tests to be added
- [x] I have added unit tests covering my code change
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly

### CODE REVIEW

**Comment by jackyzha0** [comment]
@parano ready for review!

File: docs/source/concepts.rst [hunk 1]
```diff
@@ -169,25 +169,27 @@ python class, simply use the :code:`infer_pip_packages=True` option.
...
+option did not work for your case (bug report highly appreciated), you can also specify
+the list of PyPI packages manually using the :code:`pip_packages` option:
 
 .. code-block:: python
 
   @bentoml.env(
-    pip_packages=['scikit-learn']
+    pip_packages=[
+      'scikit-learn',
+      'pandas @https://github.com/pypa/pip/archive/1.3.1.zip',
```
Comment (parano) [inline docs/source/concepts.rst:180]: 👍

File: bentoml/service/env.py [hunk 2]
```diff
@@ -323,29 +322,42 @@ def save(self, path):
         # Parse pip packages from specified requirements.txt file, note that this file
```
Comment (parano) [inline bentoml/service/env.py:322]: remove the comment here?

File: bentoml/service/env.py [hunk 3]
```diff
@@ -323,29 +322,42 @@ def save(self, path):
...
-            if self._pip_extra_index_url:
-                f.write(
-                    f"--extra-index-url={self._pip_extra_index_url}\n".encode("utf-8")
-                )
-            pip_content = '\n'.join(
-                [str(pkg_req) for pkg_req in self._pip_packages.values()]
-            ).encode("utf-8")
-            f.write(pip_content)
+        if self._requirements_txt_file:
+            logger.info(
```
Comment (parano) [inline bentoml/service/env.py:336]: remove this logging here and show a warning in the `@env` constructor instead? that way we can give the user more specific warning and help message

Comment (parano) [reply bentoml/service/env.py:336]: in `env_decorator`, we can have something like this:

```python
if requirements_txt_file:
    if pip_packages:
         logger.warn(...)
    if infer_pip_packages is not None:
         logger.warn(...)
    if pip_index_url or pip_trusted_host or pip_extra_index_url:
         # tell user to put those options in the requirements txt file directly instead
         logger.warn(...)
```

File: bentoml/service/env.py [hunk 4]
```diff
@@ -367,9 +379,10 @@ def to_dict(self):
                 str(pkg_req) for pkg_req in self._pip_packages.values()
             ]
 
-        env_dict["conda_env"] = self._conda_env._conda_env
+        if self._requirements_txt_file:
+            env_dict["requirements_txt"] = self._requirements_txt_file
```
Comment (parano) [inline bentoml/service/env.py:383]: as we discussed earlier about this metadata field, it should probably read the file content here?

File: tests/test_auto_pip_dependency.py [hunk 5]
```diff
@@ -20,11 +24,34 @@ def test_auto_adapter_dependencies(bento_bundle_path):
...
+
+def _parse_dependencies(path):
+    with open(os.path.join(path, 'requirements.txt')) as f:
+        requirements_txt_content = f.read()
+
+    return requirements_txt_content.split('\n')
+
+
+def _dependencies_to_requirements(deps):
+    print(deps)
```
Comment (parano) [inline tests/test_auto_pip_dependency.py:48]: remove print?

File: tests/test_auto_pip_dependency.py [hunk 6]
```diff
@@ -35,20 +62,37 @@ def test_auto_artifact_dependencies():
...
+    saved_path = iris_classifier_service.save()
+
+    dependencies = _dependencies_to_requirements(_parse_dependencies(saved_path))
+    _assert_in_dependencies(
+        ['scikit-learn', 'azure-cli', 'psycopg2-binary', 'bentoml'], dependencies
+    )
+
+    bs_metadata = load_bento_service_metadata(saved_path)
+    requirements_txt = bs_metadata.env.requirements_txt
+    assert requirements_txt == "./tests/pipenv_requirements.txt"
```
Comment (parano) [inline tests/test_auto_pip_dependency.py:95]: as discussed, update to be matching the content of the file

**Review by parano — CHANGES_REQUESTED** [review]
finished a pass with a few minor comments, overall looks really good!

File: bentoml/service/env.py (hunk 3 above)
Comment (jackyzha0) [reply bentoml/service/env.py:336]: why would we want to warn on `infer_pip_packages is not None`?

File: bentoml/service/__init__.py [hunk 7]
```diff
@@ -248,16 +248,24 @@ def env_decorator(
...
-            "`infer_pip_packages` instead, e.g. `@env(infer_pip_packages=True)`"
-        )
+    if requirements_txt_file:
+        if pip_packages:
+            logger.warning("Ignoring pip_packages as requirements_txt_file is set.")
+        if pip_index_url or pip_trusted_host or pip_extra_index_url:
+            logger.warning(
+                "Ignoring pip related options as requirements_txt_file is set."
+            )
+    else:
```
Comment (parano) [inline bentoml/service/__init__.py:258]: why this `else`  here? I think those warnings are orthogonal to the `requirements_txt_file` related warnings above right?

File: bentoml/service/env.py (hunk 3 above)
Comment (parano) [reply bentoml/service/env.py]: @jackyzha0  because `infer_pip_packages` controls if BentoML tries to add pip packages to the `self.packages` list, which will be ignored if `requirements_txt_file` is provided

**Comment by parano** [comment]
Currently, in the YataiService web UI, the page only shows the `pip_packages` proto field. it would be great if you could update that to show the content of `requirements.txt` if it is not empty. Maybe we can do that in a follow-up PR.
